File: python/fan-out/url-check/handler_test.py
from io import BytesIO
from types import SimpleNamespace

import requests

try:
    from . import handler as h
except ImportError:
    import handler as h


class FakeResponse:
    def __init__(self, status_code, content_type="text/plain"):
        self.status_code = status_code
        self.headers = {"Content-Type": content_type}
        self.raw = BytesIO(b"response body")
        self.closed = False

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.closed = True


def event(target):
    return SimpleNamespace(body=target.encode())


def test_reports_healthy_url(monkeypatch):
    fetched = FakeResponse(204)
    monkeypatch.setattr(h.requests, "get", lambda *args, **kwargs: fetched)
    response = h.handle(event("https://www.openfaas.com/"), {})

    assert response["statusCode"] == 200
    assert response["body"]["reachable"] is True
    assert response["body"]["healthy"] is True
    assert fetched.closed is True


def test_reports_unhealthy_status(monkeypatch):
    monkeypatch.setattr(h.requests, "get", lambda *args, **kwargs: FakeResponse(503))
    response = h.handle(event("https://one.example"), {})

    assert response["body"]["reachable"] is True
    assert response["body"]["healthy"] is False


def test_reports_network_error(monkeypatch):
    def fail(*args, **kwargs):
        raise requests.Timeout("request timed out")

    monkeypatch.setattr(h.requests, "get", fail)
    response = h.handle(event("https://one.example"), {})

    assert response["statusCode"] == 200
    assert response["body"]["reachable"] is False
    assert "timed out" in response["body"]["error"]


def test_rejects_invalid_url():
    response = h.handle(event("file:///etc/passwd"), {})
    assert response["statusCode"] == 400

File: node/fan-out/url-check/handler.test.js
'use strict'

const assert = require('node:assert/strict')
const { afterEach, test } = require('node:test')

const handler = require('./handler')

const originalFetch = global.fetch

afterEach(() => {
  global.fetch = originalFetch
})

test('reports a healthy URL', async () => {
  global.fetch = async () => new Response(null, {
    status: 204,
    headers: { 'Content-Type': 'text/plain' }
  })

  const context = new TestContext()
  await handler({ body: 'https://www.openfaas.com/' }, context)

  assert.equal(context.statusCode, 200)
  assert.equal(context.body.reachable, true)
  assert.equal(context.body.healthy, true)
})

test('reports an unhealthy HTTP status', async () => {
  global.fetch = async () => new Response('unavailable', { status: 503 })

  const context = new TestContext()
  await handler({ body: 'https://one.example' }, context)

  assert.equal(context.body.reachable, true)
  assert.equal(context.body.healthy, false)
  assert.equal(context.body.status_code, 503)
})

test('returns network failures as a health result', async () => {
  global.fetch = async () => {
    throw new Error('request timed out')
  }

  const context = new TestContext()
  await handler({ body: 'https://one.example' }, context)

  assert.equal(context.statusCode, 200)
  assert.equal(context.body.reachable, false)
  assert.match(context.body.error, /timed out/)
})

test('rejects a non-HTTP URL', async () => {
  const context = new TestContext()
  await handler({ body: 'file:///etc/passwd' }, context)

  assert.equal(context.statusCode, 400)
})

class TestContext {
  status (value) {
    this.statusCode = value
    return this
  }

  headers (value) {
    this.headerValues = value
    return this
  }

  succeed (value) {
    this.body = value
    return value
  }
}
